ddprof-lib/src/main/cpp/flightRecorder.cpp

```diff
@@ -110,27 +115,81 @@ void Lookup::cutArguments(char* func) {
+
+    jvmti->GetPhase(&phase);
+    if ((phase & (JVMTI_PHASE_START | JVMTI_PHASE_LIVE)) == 0) {
+        return;
     }
-    fillMethodInfo(mi, cmi_ptr ? cmi_ptr : _rec->_method_cache.emptyInfo());
-}
 
-void Lookup::fillMethodInfo(MethodInfo* mi, std::shared_ptr<AbstractMethodInfo> cmi_ptr) {
-    AbstractMethodInfo* cmi = cmi_ptr.get();
-    std::string class_name = cmi->className();
-    std::string method_name = cmi->methodName();
-    std::string method_sig = cmi->methodSignature();
+    bool entry = false;
+    bool deallocateStrings = false;
+    if (VMMethod::check_jmethodID(method) && jvmti->GetMethodDeclaringClass(method, &method_class) == 0 &&
+            jvmti->GetClassSignature(method_class, &class_name, NULL) == 0 &&
+            jvmti->GetMethodName(method, &method_name, &method_sig, NULL) == 0) {
+
+        if (jvmti->GetClassModifiers(method_class, &class_modifiers) == 0 && jvmti->GetMethodModifiers(method, &modifiers) == 0) {
+            // class constants are written without the modifiers info
+            // in order to be able to identify 'hidden' frames the "SYNTHETIC" and "BRIDGE" modifiers will be propagated to methods
+            if (class_modifiers & 0x1000) {
+                modifiers |= 0x1000;
+            }
+            if (class_modifiers & 0x0040) {
+                modifiers |= 0x0040;
+            }
+        }
+
+        jvmti->GetLineNumberTable(method, &line_number_table_size, &line_number_table);
 
-    const char* klass_name = class_name.c_str();
-    mi->_class = _classes->lookup(klass_name + 1, strlen(klass_name) - 2);
-    mi->_name = _symbols.lookup(method_name.c_str());
-    mi->_sig = _symbols.lookup(method_sig.c_str());
-    mi->_modifiers = cmi->_modifiers;
-    mi->_line_number_table = cmi->_line_number_table;
-    mi->_type = cmi->_type;
-    mi->_is_entry = cmi->isEntry();
+        // Check if the frame is Thread.run or inherits from it
+        if (strncmp(method_name, "run", 3) == 0 && strncmp(method_sig, "()V", 3) == 0) {
+            jclass Thread_class = jni->FindClass("java/lang/Thread");
+            jmethodID equals =
+                jni->GetMethodID(jni->FindClass("java/lang/Class"), "equals", "(Ljava/lang/Object;)Z");
+            jclass klass = method_class;
+            do {
+                if (jni->CallBooleanMethod(Thread_class, equals, klass)) {
+                    entry = true;
+                    break;
+                }
+            } while ((klass = jni->GetSuperclass(klass)) != NULL);
+        } else if ((modifiers & 9) != 0 && strncmp(method_name, "main", 4) == 0 && strncmp(method_sig, "(Ljava/lang/String;)V", 21)) {
+            // public static void main(String[] args) - 'public static' translates to modifier bits 0 and 3, hence check for '9'
+            entry = true;
+        }
+        deallocateStrings = true;
+    } else {
+        class_name = (char*)"L;";
+        method_name = (char*)"jvmtiError";
+        method_sig = (char*)"()L;";
```

Comment:
AFAIK, this is safe. At least, the same code has been there since beginning.

Reply:
I would say this is a style discussion.
The type is here to help. You would usually avoid c-style casts in favour of the appropriate c++ cast. And you usually would avoid using a const_cast. I think you introduced this to unify the code paths. I probably would have added an extra variable to avoid this.